### PR TITLE
Gaussian mixture model implementation

### PR DESCRIPTION
**Pull request by salvisolamartinell**
Distributed implementation of the Gaussian mixture model with methods fit, fit_predict and predict. Allows the same options as scikit-learn's GaussianMixture, except `n_init` (not implemented, only one initialization is done for launch) and `cov_type` (only `full` is supported for now).

### CODE REVIEW

File: dislib/cluster/gm/base.py [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+    total = 0
+    count = 0
+    for sum_s, count_s in subset_tuples:
+        total += sum_s
+        count += count_s
+    return total/count
+
+
+@task(returns=2)
+def _estimate_responsabilities(subset, weights, means, precisions_cholesky,
```
Comment (javicid) [inline dislib/cluster/gm/base.py]: There is a typo in responsibilities

File: dislib/cluster/gm/base.py [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+        precisions_chol = 1. / np.sqrt(covariances)
+    return precisions_chol
+
+
+@task(returns=1)
+def _mean_reduce(*subset_tuples):
+    total = 0
+    count = 0
+    for sum_s, count_s in subset_tuples:
+        total += sum_s
```
[the remark this answers is not in the document]
Comment (salvisolamartinell) [reply dislib/cluster/gm/base.py]: I timed different options in ipython, the fastest are `total, count = [sum(x) for x in zip(*tuples)]` and `total, count = map(sum, zip(*tuples))`, although the difference with the current version is not big. Using numpy is slower, as it implicitly creates an array. I'm going to use `total, count = map(sum, zip(*tuples))`.
```
import numpy as np  

def option_1(*tuples):
    total = 0
    count = 0
    for sum_s, count_s in tuples:
        total += sum_s
        count += count_s
    return total/count

def option_2(*tuples):
    total = sum(t[0] for t in tuples)
    count = sum(t[1] for t in tuples)
    return total/count

def option_3(*tuples):
    total, count = [sum(x) for x in zip(*tuples)]
    return total/count

def option_4(*tuples):
    total, count = map(sum, zip(*tuples))
    return total/count  

def option_2_np(*tuples):
    total = np.sum(np.fromiter((t[0] for t in tuples), dtype=float, count=len(tuples)))
    count = np.sum(np.fromiter((t[1] for t in tuples), dtype=int, count=len(tuples)))
    return total/count

def option_3_np(*tuples):
    total, count = [np.sum(x) for x in zip(*tuples)]
    return total/count

def option_4_np(*tuples):
    total, count = map(np.sum, zip(*tuples))
    return total/count

tuples_A = [(1.23, 30), (2.24, 30), (3.25, 30), (4.26, 30), (5.27, 20)]*10
tuples_B = [(1.23, 3000), (2.24, 3000), (3.25, 3000), (4.26, 3000), (5.27, 2000)]*1000

%timeit option_1(*tuples_A)
%timeit option_2(*tuples_A)
%timeit option_3(*tuples_A)
%timeit option_4(*tuples_A)
%timeit option_2_np(*tuples_A)
%timeit option_3_np(*tuples_A)
%timeit option_4_np(*tuples_A)
%timeit option_1(*tuples_B)
%timeit option_2(*tuples_B)
%timeit option_3(*tuples_B)
%timeit option_4(*tuples_B)
%timeit option_2_np(*tuples_B)
%timeit option_3_np(*tuples_B)
%timeit option_4_np(*tuples_B)
```

File: dislib/cluster/gm/base.py [hunk 3]
```diff
@@ -0,0 +1,716 @@
...
+
+from pycompss.api.task import task
+
+from dislib.cluster import KMeans
+from dislib.data import Dataset, Subset
+
+
+def _estimate_parameters(dataset, resp):
+    """Estimate the Gaussian distribution weights and means.
+
```
[the remark this answers is not in the document]
Comment (salvisolamartinell) [reply dislib/cluster/gm/base.py]: Ok, I just reused scikit-learn explanations.

File: dislib/cluster/gm/base.py [hunk 4]
```diff
@@ -0,0 +1,716 @@
...
+        Parameters
+        ----------
+        dataset : dislib.data.Dataset
+            Data points.
+        Returns
+        -------
+        self
+        """
+        self.fit_predict(dataset)
+        return self
```
[the remark this answers is not in the document]
Comment (salvisolamartinell) [reply dislib/cluster/gm/base.py]: I guess it's for being able to do concatenated calls to gm methods, e.g., gm.fit(data1).predict(data2). But it's not necessary, so I will remove it to make all interfaces of this library equivalent.

File: dislib/cluster/gm/base.py (hunk 1 above)
Comment (salvisolamartinell) [reply dislib/cluster/gm/base.py]: Corrected.

File: dislib/cluster/gm/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (javicid) [reply dislib/cluster/gm/base.py]: Great

**Comment by salvisolamartinell** [comment]
> Nice work! I have run clustering_comparison and the algorithm is not performing the right clustering with the aniso data. Do you know why this is?
I've been comparing with sklearn. Both initialize with 'kmeans', but sklearn returned much more stable initial points. I thought this was due to sklearn kmeans using max_iter=300 and init='k-means++', but even when this was equalized sklearn kmeans was more stable.

Also I tried to compare initializing with 'random', and the results look similar for dislib and sklearn, but the algorithms seldom converge (probably with different params and multiple initializations the results would be better).

I can change kmeans max_iter from 10 to 300, but it isn't enough. We should review how to make our KMeans equivalent to sklearn's.

> Also, can you add a short description for each new test as discussed in #78 ?
Done.

**Comment by javicid** [comment]
Are you saying that GMM produces different results than sklearn because of K-Means? 

I assume you are passing a random_state to K-Means.

**Comment by salvisolamartinell** [comment]
There was an error in K-means: default tolerance was set as `tol=1 ** -4` (that is tol=1) while `tol=1e-4` was documented. I fixed this error and now the GaussianMixture works well in clustering_comparison with the aniso dataset. Note that sklearn K-means still works with different default parameters (`init='k-means++'`, `n_init=10`, `max_iter=300`, `algorithm='auto'`), but these differences do not affect the clustering_comparison.

**Comment by javicid** [comment]
Nice!

**Pull request merged**